File: src/main/java/no/nav/syfo/kafka/KafkaErrorHandler.java
package no.nav.syfo.kafka;

import io.micrometer.core.instrument.MeterRegistry;
import io.micrometer.core.instrument.Tags;
import lombok.extern.slf4j.Slf4j;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.common.errors.TopicAuthorizationException;
import org.springframework.kafka.listener.ContainerAwareErrorHandler;
import org.springframework.kafka.listener.ContainerStoppingErrorHandler;
import org.springframework.kafka.listener.MessageListenerContainer;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
@Slf4j
public class KafkaErrorHandler implements ContainerAwareErrorHandler {

    private static ContainerStoppingErrorHandler STOPPING_ERROR_HANDLER = new ContainerStoppingErrorHandler();

    private MeterRegistry registry;

    public KafkaErrorHandler(MeterRegistry registry) {
        this.registry = registry;
    }

    @Override
    public void handle(Exception thrownException, List<ConsumerRecord<?, ?>> records, Consumer<?, ?> consumer, MessageListenerContainer container) {
        log.error("Feil i listener:", thrownException);

        /*
            Dette er en hack for å redde Kafka fra en evig løkke om koblingen mot AD går ned. Da vil vi inkrementere en feilmetrikk
            som etter en del gjentakelser vil føre til at isAlive flippes og poden blir restartet av kubernetes.
         */
        if (exceptionIsClass(thrownException, TopicAuthorizationException.class)) {
            log.error("Kafka infrastrukturfeil. TopicAuthorizationException ved lesing av topic");
            registry.counter("$APPLIKASJONSNAVN$.kafka.feil", Tags.of("type", "fatale")).increment();
            return;
        }

        records
                .forEach(record -> log.error("Feil i prossesseringen av record med offset:{} og innhold:{}", record.offset(), record.value()));

        STOPPING_ERROR_HANDLER.handle(thrownException, records, consumer, container);
    }

    private boolean exceptionIsClass(Throwable t, Class klazz) {
        int maxdepth = 10;
        while (maxdepth-- > 0 && t != null && !klazz.isInstance(t)) {
            t = t.getCause();
        }

        return klazz.isInstance(t);
    }
}

File: src/main/kotlin/no/nav/syfo/kafka/KafkaErrorHandler.kt
package no.nav.syfo.kafka

import no.nav.syfo.logger
import org.apache.kafka.clients.consumer.Consumer
import org.apache.kafka.clients.consumer.ConsumerRecord
import org.apache.kafka.clients.consumer.ConsumerRecords
import org.springframework.kafka.listener.*
import org.springframework.stereotype.Component
import org.springframework.util.backoff.ExponentialBackOff
import java.lang.Exception

@Component
class KafkaErrorHandler : DefaultErrorHandler(
    ExponentialBackOff(1000L, 1.5).apply {
        // 8 minutter, som er mindre enn max.poll.interval.ms på 10 minutter.
        maxInterval = 60_000L * 8
    }
) {

    val log = logger()

    override fun handleRemaining(
        thrownException: java.lang.Exception,
        records: MutableList<ConsumerRecord<*, *>>,
        consumer: Consumer<*, *>,
        container: MessageListenerContainer
    ) {
        records.forEach { record ->
            log.error(
                "Feil i prossessering av record med offset: ${record.offset()}, key: ${record.key()} på topic ${record.topic()}",
                thrownException
            )
        }
        if (records.isEmpty()) {
            log.error("Feil i listener uten noen records", thrownException)
        }
        super.handleRemaining(thrownException, records, consumer, container)
    }

    override fun handleBatch(
        thrownException: Exception,
        records: ConsumerRecords<*, *>,
        consumer: Consumer<*, *>,
        container: MessageListenerContainer,
        invokeListener: Runnable
    ) {
        records.forEach { record ->
            log.error(
                "Feil i prossessering av record med offset: ${record.offset()}, key: ${record.key()} på topic ${record.topic()}",
                thrownException
            )
        }
        if (records.isEmpty()) {
            log.error("Feil i listener uten noen records", thrownException)
        }
        super.handleBatch(thrownException, records, consumer, container, invokeListener)
    }
}
